postal/internal/dependency_mirror.go

```diff
@@ -0,0 +1,143 @@
+package internal
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+)
+
+type DependencyMirrorResolver struct {
+	bindingResolver BindingResolver
+}
+
+func NewDependencyMirrorResolver(bindingResolver BindingResolver) DependencyMirrorResolver {
+	return DependencyMirrorResolver{
+		bindingResolver: bindingResolver,
+	}
+}
+
+func formatAndVerifyMirror(mirror, uri string) (string, error) {
+	mirrorURL, err := url.Parse(mirror)
+	if err != nil {
+		return "", err
+	}
+
+	uriURL, err := url.Parse(uri)
+	if err != nil {
+		return "", err
+	}
+
+	if mirrorURL.Scheme != "https" && mirrorURL.Scheme != "file" {
+		return "", fmt.Errorf("invalid mirror scheme")
+	}
+
+	mirrorURL.Path = strings.Replace(mirrorURL.Path, "{originalHost}", uriURL.Host+uriURL.Path, 1)
+	return mirrorURL.String(), nil
+}
+
+func (d DependencyMirrorResolver) FindDependencyMirror(uri, platformDir string) (string, error) {
+	mirror, err := d.findMirrorFromEnv(uri)
+	if err != nil {
+		return "", err
+	}
+
+	if mirror != "" {
+		return formatAndVerifyMirror(mirror, uri)
+	}
+
+	mirror, err = d.findMirrorFromBinding(uri, platformDir)
+	if err != nil {
+		return "", err
+	}
+
+	if mirror != "" {
+		return formatAndVerifyMirror(mirror, uri)
+	}
+
+	return "", nil
+}
+
+func (d DependencyMirrorResolver) findMirrorFromEnv(uri string) (string, error) {
+	const DefaultMirror = "BP_DEPENDENCY_MIRROR"
+	const DefaultMirrorPrefix = "BP_DEPENDENCY_MIRROR_"
+	mirrors := make(map[string]string)
+	environmentVariables := os.Environ()
+	for _, ev := range environmentVariables {
+		idx := strings.Index(ev, "=")
+		if idx < 0 {
+			continue
+		}
+
+		key := ev[:idx]
+		value := ev[idx+1:]
+		if !strings.Contains(key, DefaultMirror) {
+			continue
+		}
+
+		if key == DefaultMirror {
+			mirrors["default"] = value
+			continue
+		}
+
+		// convert key
+		hostnameIdx := len(DefaultMirrorPrefix)
+		hostname := key[hostnameIdx:]
+		hostname = strings.ReplaceAll(hostname, "_", ".")
+		hostname = strings.ReplaceAll(hostname, "__", "-")
```

Comment:
This logic doesn't work perfectly, might need to add a test for it.

For example, if the input is `BP_DEPENDENCY_MIRROR_TESTING_123__ABC`, the expected output is `testing.123-abc`, but this code actually outputs `testing.123..abc`

I think these lines need to get reversed, so double underscores are converted first